source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
+            const auto &cell_status = std::get<1>(*cit_qcrel);
+            const auto &dealii_cell = std::get<2>(*cit_qcrel);
+
+#  ifdef DEBUG
+            // Assertions about the tree structure.
+            switch (cell_status)
+              {
+                case parallel::distributed::Triangulation<dim, spacedim>::
+                  CELL_PERSIST:
+                case parallel::distributed::Triangulation<dim, spacedim>::
+                  CELL_REFINE:
+                  // double check the condition that we will only ever attach
+                  // data to active cells when we get here
+                  Assert(dealii_cell->active(), ExcInternalError());
+                  break;
+
+                case parallel::distributed::Triangulation<dim, spacedim>::
+                  CELL_COARSEN:
+                  // double check the condition that we will only ever attach
+                  // data to cells with children when we get here. however, we
+                  // can only tolerate one level of coarsening at a time, so
+                  // check that the children are all active
+                  Assert(dealii_cell->active() == false, ExcInternalError());
+                  for (unsigned int c = 0;
+                       c < GeometryInfo<dim>::max_children_per_cell;
+                       ++c)
+                    Assert(dealii_cell->child(c)->active(), ExcInternalError());
+                  break;
+
+                case parallel::distributed::Triangulation<dim, spacedim>::
+                  CELL_INVALID:
+                  // do nothing on invalid cells
+                  break;
+
+                default:
+                  Assert(false, ExcInternalError());
+                  break;
+              }
+#  endif
+
+            // Reserve memory corresponding to the amount of callback
+            // function that will be called.
+            // On cells flagged with CELL_INVALID, only its CellStatus
+            // will be stored.
+            const unsigned int num_callbacks_on_cell =
+              1 +
+              ((cell_status ==
+                parallel::distributed::Triangulation<dim,
+                                                     spacedim>::CELL_INVALID) ?
+                 0 :
+                 pack_callbacks.size());
+            it_data_cell->resize(num_callbacks_on_cell);
+
+            // We continue with packing all data on this specific cell.
+            // First, we pack the CellStatus information.
+            auto it_data_callback = it_data_cell->begin();
+            *it_data_callback =
+              Utilities::pack(cell_status, /*allow_compression=*/false);
+
+            // Proceed with all registered callback functions.
+            // Skip cells with the CELL_INVALID flag.
+            if (cell_status !=
+                parallel::distributed::Triangulation<dim,
+                                                     spacedim>::CELL_INVALID)
+              {
+                ++it_data_callback;
```

Comment:
I would move this immediately after the place above where you write into it, i.e., below the call to `Utilities::pack(...)`